sysvabi64/sysvabi64.rst

```diff
@@ -720,22 +724,37 @@ open-source compilers.
 
 .. table:: Code Model Support
 
-  +-------+----------------+----------------+
-  | Code  | GCC 11         | Clang 12.0     |
-  | Model |                |                |
-  +=======+================+================+
-  | tiny  | yes            | yes            |
-  |       |                |                |
-  +-------+----------------+----------------+
-  | small | pic and PIC    | pic and PIC    |
-  |       | implemented    | accepted but   |
-  |       |                | PIC used       |
-  +-------+----------------+----------------+
-  | large | no-pic support | no-pic support |
-  |       |                |                |
-  |       |                |                |
-  |       |                |                |
-  +-------+----------------+----------------+
+  +--------+----------------+----------------+
+  | Code   | GCC 11         | Clang 12.0     |
+  | Model  |                |                |
+  +========+================+================+
+  | tiny   | yes            | yes            |
+  |        |                |                |
+  +--------+----------------+----------------+
+  | small  | pic and PIC    | pic and PIC    |
+  |        | implemented    | accepted but   |
+  |        |                | PIC used       |
+  +--------+----------------+----------------+
+  | medium | pic and PIC    | pic and PIC    |
+  |        | proposed       | proposed       |
+  |        |                |                |
+  +--------+----------------+----------------+
+  | large  | no-pic support | no-pic support |
+  |        |                |                |
+  |        |                |                |
+  |        |                |                |
+  +--------+----------------+----------------+
+
+
+Medium code model
```

Comment:
Here's an alternative wording suggestion.

The medium code model is suitable for programs that do not have large amounts of executable code, but do contain large static data objects such as large look up tables. The medium code model separates data into ``small data`` and ``large data``. The ``small data`` is addressed in the same way as the small code model, with the same maximum size restrictions as the small code model. The ``large data`` is addressed via the GOT and if placed after the ``small data`` does not have a maximum size limit.

Data is defined as ``large data`` if it is not Relocation Read-Only (RELRO) and any of the following are true:
* The data is >= 64 KiB in size.
* The data is of unknown size.
* The data is represented as a symbol with type ``STT_COMMON``.

All other data is ``small data``.

``large data`` are placed in sections with a prefix of ``l``, for example ``.lbss``, ``.ldata`` and ``.lrodata``. At static link time the ``large data`` sections must be placed after all ``small data`` sections. There is no large data equivalent of the ``.data.rel.ro`` section as all Relocation Read-Only data is ``small data``.

Reply:
> There is no large data equivalent of the `.data.rel.ro` section as all Relocation Read-Only data is `small data`.

Is there a reason behind this? It's easy to create a large rel.ro symbol, and for the medium model to work correctly, it has to be treated like large data and placed at the end.

Reply:
There's no fundamental reason, although IIUC the only reason to use data.rel.ro is if data contains pointers that need static initialisation. How many large objects, or arrays of objects contains that kind of data.

Reply:
It should be very rare, but it seems best to avoid odd limitations that make it less generic.

Reply:
glibc/FreeBSD/musl's ld.so implementations only support one PT_GNU_RELRO.

LLD therefore has a diagnostic `<section_name> is not contiguous with other relro sections`